src/lxml/sax.py

```diff
@@ -227,21 +235,35 @@ def _recursive_saxify(self, element, prefixes):
         if element.text:
             content_handler.characters(element.text)
         for child in element:
-            self._recursive_saxify(child, prefixes)
+            self._recursive_saxify(child)
         content_handler.endElementNS((ns_uri, local_name), qname)
         for prefix, uri in new_prefixes:
             content_handler.endPrefixMapping(prefix)
         if element.tail:
             content_handler.characters(element.tail)
 
-    def _build_qname(self, ns_uri, local_name, prefixes, new_prefixes):
+    def _build_qname(self, ns_uri, local_name, prefixes, preferred_prefix):
         if ns_uri is None:
             return local_name
-        try:
-            prefix = prefixes[ns_uri]
-        except KeyError:
-            prefix = prefixes[ns_uri] = 'ns%02d' % len(prefixes)
-            new_prefixes.append( (prefix, ns_uri) )
+
+        if prefixes.get(preferred_prefix) == ns_uri:
+            prefix = preferred_prefix
+        else:
+            # Pick the first matching prefix:
+            for pfx in sorted(prefixes, key=str):
```

Comment:
I tested that and though namespaces ended up as always being lowercase? Maybe I messed up. Anyway, that is then an implicit and unclear feature, so that's not good code, I'll change this.